### PR TITLE
FIR checker: check if callable reference targets are allowed members

### PR DESCRIPTION
**Pull request by jsjeon**
Otherwise we can report:
* CALLABLE_REFERENCE_TO_ANNOTATION_CONSTRUCTORS
* EXTENSION_IN_CLASS_REFERENCE_NOT_ALLOWED

if callable reference target is either annotation or member extension.

------

We need two meta errors:
* UNSUPPORTED
* UNSUPPORTED_FEATURE

to report cases like reference to local variables or value parameters, or empty LHS (e.g., `::member`) without the corresponding feature enabled.

------

Instead of the entire qualified expression, we need to report on the callee reference, i.e., after `::`. Hence, added a util to find the _last_ child of a certain type. This also reveals a similar LT issue where a source of a subtree needs to be rebased.

### CODE REVIEW

**Review by tgeng — COMMENTED** [review]
LGTM. Though I seem to have some trouble to build due to generated code in `KtFirDiagnosticsImpl.kt`.

**Comment by jsjeon** [comment]
Thanks for catching that. I should have paid attention to `idea` tests. Fixed type parameter issues in `KtFirDiagnostics*`.

**Review by tgeng — APPROVED** [review]

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirCallableReferenceChecker.kt [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        }
+        val referredDeclaration = reference.resolvedSymbol.fir
+        if (referredDeclaration is FirConstructor && referredDeclaration.containingClass?.classKind == ClassKind.ANNOTATION_CLASS) {
+            reporter.reportOn(referenceSource, FirErrors.CALLABLE_REFERENCE_TO_ANNOTATION_CONSTRUCTORS, context)
+        }
+        if (referredDeclaration is FirCallableMemberDeclaration<*> && referredDeclaration.isMemberExtension) {
+            reporter.reportOn(referenceSource, FirErrors.EXTENSION_IN_CLASS_REFERENCE_NOT_ALLOWED, referredDeclaration, context)
+        }
+        if (referredDeclaration is FirVariable<*> &&
+            referredDeclaration !is FirField &&
```
Comment (punzki) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirCallableReferenceChecker.kt:69]: You can have a reference to a field?

Can you try this:
```kt
val i: Int = 10
    get() {
        ::field  // Error in FE1.0: Unsupported [References to variables aren't supported yet]
        return field
    }
```

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirCallableReferenceChecker.kt [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+            (referredDeclaration !is FirProperty || referredDeclaration.isLocal)
+        ) {
+            reporter.reportOn(referenceSource, FirErrors.UNSUPPORTED, "References to variables aren't supported yet", context)
+        }
+    }
+
+    private val FirCallableMemberDeclaration<*>.isMemberExtension: Boolean
```
Comment (punzki) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirCallableReferenceChecker.kt:76]: Maybe move to `FirDeclarationUtil`? Or too limited in usefulness?

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirCallableReferenceChecker.kt (hunk 1 above)
Comment (jsjeon) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirCallableReferenceChecker.kt]: It looks like there is no such test, thanks!

Btw, it's still `FirProperty` that such backing field references to, and yet it's not `local`. The check for `!is FirField` is there to discard Java field or delegate field, but seems a little bit vague. I rewrote here to clarify what we're reporting.

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirCallableReferenceChecker.kt (hunk 2 above)
Comment (jsjeon) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirCallableReferenceChecker.kt]: Why not, moved. :)

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirCallableReferenceChecker.kt (hunk 1 above)
Comment (punzki) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirCallableReferenceChecker.kt]: Looks clearer, thanks!

**Review by punzki — APPROVED** [review]

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirCallableReferenceChecker.kt [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+    // See FE 1.0 [DoubleColonExpressionResolver#checkReferenceIsToAllowedMember]
+    private fun checkReferenceIsToAllowedMember(
+        callableReferenceAccess: FirCallableReferenceAccess,
+        context: CheckerContext,
+        reporter: DiagnosticReporter
+    ) {
+        // UNRESOLVED_REFERENCE will be reported separately.
+        if (callableReferenceAccess.calleeReference !is FirResolvedNamedReference) return
+        val reference = callableReferenceAccess.calleeReference as FirResolvedNamedReference
```
Comment (mglukhikh) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirCallableReferenceChecker.kt:44]: It's better to write `val referenace = callableReferenceAccess.calleeReference as? FirResolvedNamedReference ?: return` and remove the upper line

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirCallableReferenceChecker.kt [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+        val source = reference.source ?: return
+        if (source.kind is FirFakeSourceElementKind) return
+        // For example, rcv::member, reference source points to the entire callable reference.
+        // Note that both `rcv` and `member` would be of [REFERENCE_EXPRESSION] kind.
+        // We want to report on the last reference expression, i.e. the one after ::
+        val referenceSource = source.getLastChild(REFERENCE_EXPRESSION) ?: return
+
+        if (!context.session.languageVersionSettings.supportsFeature(LanguageFeature.CallableReferencesToClassMembersWithEmptyLHS)) {
```
Comment (mglukhikh) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirCallableReferenceChecker.kt:53]: Please don't check for the features which appeared in 1.4 or earlier. FIR anyway shouldn't support so old stuff.

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirCallableReferenceChecker.kt [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        // UNRESOLVED_REFERENCE will be reported separately.
+        if (callableReferenceAccess.calleeReference !is FirResolvedNamedReference) return
+        val reference = callableReferenceAccess.calleeReference as FirResolvedNamedReference
+
+        val source = reference.source ?: return
+        if (source.kind is FirFakeSourceElementKind) return
+        // For example, rcv::member, reference source points to the entire callable reference.
+        // Note that both `rcv` and `member` would be of [REFERENCE_EXPRESSION] kind.
+        // We want to report on the last reference expression, i.e. the one after ::
+        val referenceSource = source.getLastChild(REFERENCE_EXPRESSION) ?: return
```
Comment (mglukhikh) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirCallableReferenceChecker.kt:51]: Please don't use children manipulation. Use positioning strategy instead.

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirCallableReferenceChecker.kt (hunk 4 above)
Comment (mglukhikh) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirCallableReferenceChecker.kt:53]: (you may leave UNSUPPORTED_FEATURE itself in PR if you wish, it anyway will be usable for newer features)

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/FirSourceChild.kt [hunk 6]
```diff
@@ -36,5 +37,48 @@ private fun FirPsiSourceElement<*>.getChild(types: Set<IElementType>, index: Int
...
+    val visitor = PsiElementFinderByType(types, index, depth)
+    return visitor.findLast(psi)?.toFirPsiSourceElement()
+}
+
+private fun FirLightSourceElement.getLastChild(types: Set<IElementType>, index: Int, depth: Int): FirSourceElement? {
+    val visitor = LighterTreeElementFinderByType(treeStructure, types, index, depth)
+    return visitor.findLast(lighterASTNode)?.let { withNode(it) }
+}
+
+private fun FirLightSourceElement.withNode(newNode: LighterASTNode): FirLightSourceElement {
```
Comment (mglukhikh) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/FirSourceChild.kt:73]: All this children manipulation code should be removed from PR. BTW in https://github.com/JetBrains/kotlin/pull/4199 the similar code does not influence anything, so I'll remove it from PR before merging it.

**Review by mglukhikh — CHANGES_REQUESTED** [review]

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirDeclarationCheckerUtils.kt [hunk 7]
```diff
@@ -220,3 +220,8 @@ internal fun FirRegularClass.isInlineOrValueClass(): Boolean {
 
     return isInline || hasModifier(KtTokens.VALUE_KEYWORD)
 }
+
+internal val FirCallableMemberDeclaration<*>.isMemberExtension: Boolean
```
Comment (mglukhikh) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirDeclarationCheckerUtils.kt:224]: `isExtensionMember` seems more clear name here.

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirCallableReferenceChecker.kt (hunk 3 above)
Comment (jsjeon) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirCallableReferenceChecker.kt]: 🤦  Not sure why I didn't do that in the first place.

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirCallableReferenceChecker.kt (hunk 4 above)
Comment (jsjeon) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirCallableReferenceChecker.kt]: Okay, removed this check, but I'll leave `UNSUPPORTED_FEATURE`, which helped me figure out an issue in IDE generator.

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirDeclarationCheckerUtils.kt (hunk 7 above)
Comment (jsjeon) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirDeclarationCheckerUtils.kt]: Removed

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/FirSourceChild.kt (hunk 6 above)
Comment (jsjeon) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/FirSourceChild.kt]: Removed, but I wonder if `UNSUPPORTED` still needs this. So, similar to wrong modifier, source of `UNSUPPORTED` depends on error kinds, and thus we can't set positioning strategy for that. Instead, we may need to find the specific source, e.g., last child for callable reference expression. I left a TODO there.

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirCallableReferenceChecker.kt (hunk 5 above)
Comment (jsjeon) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirCallableReferenceChecker.kt]: Changed to use `REFERENCE_BY_QUALIFIED` positioning strategy, and found what that strategy missed: see the 2nd commit that fixed positions for callable references, resulting in many `// FIR_IDENTICAL`

**Review by mglukhikh — APPROVED** [review]

**Comment by mglukhikh** [comment]
I've got the following errors in full pipeline test
```
cli.main: main-java-production, 67 files (8083 lines)
        1st error: 'useJavac' is a member and an extension at the same time. References to such elements are not allowed
jps-plugin.main: main-java-production, 45 files (5089 lines)
        1st error: 'target' is a member and an extension at the same time. References to such elements are not allowed
idea.main: main-java-production, 1297 files (147212 lines)
        1st error: 'leadingIndent' is a member and an extension at the same time. References to such elements are not allowed
idea-fir.main: main-java-production, 21 files (1597 lines)
        1st error: 'hasSuitableExtensionReceiver' is a member and an extension at the same time. References to such elements are not allowed
idea-new-project-wizard.main: main-java-production, 56 files (4369 lines)
        1st error: 'find' is a member and an extension at the same time. References to such elements are not allowed
```

**Comment by jsjeon** [comment]
> I've got the following errors in full pipeline test
> 
> ```
> cli.main: main-java-production, 67 files (8083 lines)
>         1st error: 'useJavac' is a member and an extension at the same time. References to such elements are not allowed
> jps-plugin.main: main-java-production, 45 files (5089 lines)
>         1st error: 'target' is a member and an extension at the same time. References to such elements are not allowed
> idea.main: main-java-production, 1297 files (147212 lines)
>         1st error: 'leadingIndent' is a member and an extension at the same time. References to such elements are not allowed
> idea-fir.main: main-java-production, 21 files (1597 lines)
>         1st error: 'hasSuitableExtensionReceiver' is a member and an extension at the same time. References to such elements are not allowed
> idea-new-project-wizard.main: main-java-production, 56 files (4369 lines)
>         1st error: 'find' is a member and an extension at the same time. References to such elements are not allowed
> ```

Ah, so we should report that on _non-local_ extension member. Updated accordingly, and double-checked full pipeline test passed.

**Comment by mglukhikh** [comment]
Merged manually 👍

**Pull request closed without merging**